File: Go/calc_ip.go
package main

import (
	"fmt"
	"math/big"
	"strings"
)

type CalcIp struct {
	ip              *big.Int
	subnet          *big.Int
	cidr            int
	networkAddress  *big.Int
	broadcastAddress *big.Int
	hostAddress     *big.Int
}

func NewCalcIp(ip string, subnet string) *CalcIp {
	c := &CalcIp{}
	if subnet == "" {
		cidr := strings.Split(ip, "/")[1]
		c.subnet = c.parseSubnetFromCidr(cidr)
		c.cidr = int(c.subnet.BitLen())
		ipStr := strings.Split(ip, "/")[0]
		c.ip = c.parseIp(ipStr)
	} else {
		c.subnet = c.parseSubnet(subnet)
		c.ip = c.parseIp(ip)
	}

	c.networkAddress = new(big.Int).And(c.ip, c.subnet)
	c.hostAddress = new(big.Int).Xor(c.subnet, big.NewInt(4294967295))
	c.broadcastAddress = new(big.Int).Or(c.networkAddress, c.hostAddress)

	return c
}

func (c *CalcIp) parseSubnet(subnet string) *big.Int {
	return c.parseIp(subnet)
}

func (c *CalcIp) parseSubnetFromCidr(cidr string) *big.Int {
	cidrInt := new(big.Int)
	cidrInt.SetString(cidr, 10)
	return new(big.Int).Lsh(big.NewInt(1), uint(32-cidrInt.Int64()))
}

func (c *CalcIp) parseIp(ip string) *big.Int {
	parts := strings.Split(ip, ".")
	ipInt := big.NewInt(0)
	for i, part := range parts {
		partInt := new(big.Int)
		partInt.SetString(part, 10)
		ipInt.Or(ipInt, new(big.Int).Lsh(partInt, uint(8*(3-i))))
	}
	return ipInt
}

func (c *CalcIp) IpString() string {
	return c.addToDottedDecimalNotation(c.ip)
}

func (c *CalcIp) SubnetString() string {
	return c.addToDottedDecimalNotation(c.subnet)
}

func (c *CalcIp) NetworkAddressString() string {
	return c.addToDottedDecimalNotation(c.networkAddress)
}

func (c *CalcIp) addToDottedDecimalNotation(ip *big.Int) string {
	ipString := ""
	for i := 3; i >= 0; i-- {
		part := new(big.Int).Rsh(ip, uint(8*i))
		ipString += part.String()
		if i > 0 {
			ipString += "."
		}
	}
	return ipString
}

File: Rust/src/calc_ip.rs
pub struct CalcIp {
    ip: u32,
    subnet: u32,
    cidr: Option<u8>,
    network_address: u32,
    broadcast_address: u32,
    host_address: u32,
}

impl CalcIp {
    pub fn new(ip: &str, subnet: Option<&str>) -> Self {
        let (ip, subnet, cidr) = if let Some(subnet) = subnet {
            (Self::parse_ip(ip), Self::parse_subnet(subnet), None)
        } else {
            let parts: Vec<&str> = ip.split('/').collect();
            let ip = Self::parse_ip(parts[0]);
            let cidr = parts[1].parse::<u8>().unwrap();
            let subnet = Self::parse_subnet_from_cidr(cidr);
            (ip, subnet, Some(cidr))
        };

        let network_address = ip & subnet;
        let host_address = !subnet;
        let broadcast_address = network_address | host_address;

        Self {
            ip,
            subnet,
            cidr,
            network_address,
            broadcast_address,
            host_address,
        }
    }

    pub fn ip_string(&self) -> String {
        Self::add_to_dotted_decimal_notation(self.ip)
    }

    pub fn subnet_string(&self) -> String {
        Self::add_to_dotted_decimal_notation(self.subnet)
    }

    pub fn network_address_string(&self) -> String {
        Self::add_to_dotted_decimal_notation(self.network_address)
    }

    pub fn get_ip(&self) -> u32 {
        self.ip
    }

    pub fn get_subnet(&self) -> u32 {
        self.subnet
    }

    pub fn get_cidr(&self) -> Option<u8> {
        self.cidr
    }

    pub fn get_network_address(&self) -> u32 {
        self.network_address
    }

    pub fn get_broadcast_address(&self) -> u32 {
        self.broadcast_address
    }

    pub fn get_host_address(&self) -> u32 {
        self.host_address
    }

    fn parse_ip(ip: &str) -> u32 {
        ip.split('.')
            .map(|octet| octet.parse::<u32>().unwrap())
            .enumerate()
            .fold(0, |acc, (i, octet)| acc + (octet << (8 * (3 - i))))
    }

    fn parse_subnet(subnet: &str) -> u32 {
        Self::parse_ip(subnet)
    }

    fn parse_subnet_from_cidr(cidr: u8) -> u32 {
        (!0u32) << (32 - cidr)
    }

    fn add_to_dotted_decimal_notation(ip: u32) -> String {
        format!(
            "{}.{}.{}.{}",
            (ip >> 24) & 0xFF,
            (ip >> 16) & 0xFF,
            (ip >> 8) & 0xFF,
            ip & 0xFF
        )
    }

    pub fn get_bin_ip_obj(&self) -> (u32, u32, u32, u32, u32, Option<u8>) {
        (
            self.ip,
            self.subnet,
            self.network_address,
            self.broadcast_address,
            self.host_address,
            self.cidr,
        )
    }
}
